Allow a pod to be actively killed by the kubelet if it runs too long

derekwaynecarr opened the pull request:
Note: Pending agreement on approach, I will add the requisite unit test and validation code.

Change:
Add a new optional field to PodSpec, `RunningTtl`, that sets a time in seconds that a pod is allowed to run before the Kubelet will actively try to fail that pod and subsequently kill its containers.  The "Running" concept aligned with the behavior in Kubelet today by enforcing the TTL value relative to the earliest started container in the pod.  The target use case is to let a user basically time-box how long a pod can run without the need to build any additional active monitoring.  It's expected that the graceful termination period for the pod will be respected once its actually properly plumbed through the Kubelet... right now, it looks to default to 10s in all cases.

This change affects all pods independent of their RestartPolicy, so a RestartPolicy Always pod will be failed, and if its not managed by a replication controller that will create a new instance of that pod, the pod will basically be done.

@bgrant0607  - This is the follow-up to discussions we had at the last community meet-up.  Loosely related to #7856 since chosen terminology aligned with the Running pod phase.  I thought about a future use case for WaitingTtl to basically set a limit to how long we are willing to wait for a docker image pull before killing the pod to hopefully schedule to a new host, but that is for another time...

/cc @abhgupta @smarterclayton

bgrant0607 commented:
On Ttl (guidelines should be added to api-conventions.md.):
1. Ttl is an acronym, so it should be all caps, similar to DNS or TCP.
2. That said, the field would either need to accept a time.Duration, with units specified as part of the value, or must have the units in the name of the field -- Seconds in this case, such as `FooDeadlineSeconds`. 

Re. Running:
1. You want to exclude pulling the images?
2. What do you want to happen if containers crash and restart during execution? If the Kubelet crashes? If the node reboots? 
3. If we implemented a Job controller (#1624), would you want the timeout to span pods, or reset every time a pod were replaced?
4. My impression is that users find the term "Running" confusing, since the containers may not actually be running. Something with less misleading connotations is likely to replace PodPhase at some point. Perhaps Active, which, like Ready, could apply fairly generically to multiple kinds of objects.

So, ActiveDeadlineSeconds?

bgrant0607 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1065,6 +1065,46 @@ func (kl *Kubelet) cleanupOrphanedVolumes(pods []*api.Pod, runningPods []*kubeco
+					continue
+				}
+
+				startedAt := containerStatus.State.Running.StartedAt
```
Restart OnFailure pods will reset StartedAt.

derekwaynecarr replied:
I needed to verify that the Docker daemon actually did reset the StartedAt field when a restart occurred, and if it did if there was a more reliable field I could pull from.  I had debated adding a RunningAt field on the PodStatus do denote when a Pod first became Ready but wanted to avoid that if possible.  Ideally, I want to have a time-stamp that represents when the pod first started and began to actually consume compute on the Node.  For this reason, I didn't want to use the CreationTimestamp field since who knows how long it took to actually schedule, or pull the associated docker images...  any opinions?

bgrant0607 replied:
From your other response, it sounds like you don't want the clock to reset. Also, the timeout should apply to the whole pod, not just an individual container. Therefore, I recommend that we add the Active condition, as discussed in #7856, which would enable you to use its LastTransitionTime.

dchen1107 replied:
@derekwaynecarr restarting a container today doesn't reset StartedAt field today. To docker, it is a totally new container anyway since kubelet restart it.

dchen1107 replied:
Since TTL is set to pod, we should work with some creationTimeStamp of Pod on local node. We don't have it today in PodStatus, but could add it easily. Also today, kubelet cached podStatus, you can retrieve creationTimeStamp from previous podStatus for a given pod. It is at the best effort since we didn't checkpoint. In a long run, kubelet should checkpoint PodSpec, which should have proper creationTimeStamp on local node, we could use that.

derekwaynecarr commented:
I have a primary goal of limiting the amount of time a Pod can be active to a Node once the Kubelet acknowledges the Pod has been scheduled.

>> You want to exclude pulling the images?

We actually have had complaints that a Pod can appear in Pending status indefinitely if the Node does not have access to an external registry/hub (see https://github.com/openshift/origin/issues/2048).  As an insider to the system, I feel like the amount of time it takes to pull image(s) should not limit the amount of time I get to run the container(s), but as an end-user, I am not sure it matters for my use case.  In the near term, I am okay if we actually blend these two concepts, but in the future, I could see a need to distinguish how long we are willing to wait to pull an image versus how long we let a container actually run.  For example, I want to limit how long to wait for image to pull, but not limit how long the container can run seems valid.

>> What do you want to happen if containers crash and restart during execution?  If the Kubelet crashes?  If the node reboots?

If containers crash and restart, I do not want that to reset the time.  If Kubelet crashes, I am OK with rogue containers that are not stopped until Kubelet is made Healthy.  If the node reboots, I am OK if the node down time is included in the duration that is accounted against the pod.

>> If we implemented a Job controller (#1624), would you want the timeout to span pods, or reset every time a pod were replaced?

If I run a Job, I have different expectations for what it means for how long the Job can take to complete.  I want the time limit of a Job to be tied to the Job, and not the Pod(s) it produces.  In a scenario where I kick off a Job to run my payroll, I want that Job to fail if it runs for 30 minutes and was never able to actually run any pods (because the pods were never scheduled, because not enough pods ran, because I lacked quota, etc.).  So I think in a Job scenario, the end-user expectation for how long it can run is different than how long a single Pod can run.  That being said, the concepts may blend.  I may have a Job that I kick off that can be active for 12 hrs, but no single Pod kicked off by that Job should run for more than 30 minutes.  I measure the time of a Job to be related to when a controller was made active, not when a pod was made active.

>> My impression is that users find the term "Running" confusing...

I agree, and I would be lying to you if I said I did not get confused when implementing this.  Looking at getPodPhase is confusing ;-)

Upon further reflection, how do you feel about the following:

1.  Add a timestamp to PodStatus that denotes when a Kubelet acknowledges that a Pod has been scheduled to the host.
2.  Add a duration in seconds that controls how long since the Kubelet acknowledges a Pod before the Kubelet will actively attempt to kill the Pod.

So how about NodeACKTimestamp and NodeActiveDeadlineSeconds?

bgrant0607 commented:
If we added an Active condition, we wouldn't need NodeACKTimestamp. That said, I wouldn't be opposed to something like PodStatus.StartTime, following whatever consistent naming convention we settle on for times (https://github.com/GoogleCloudPlatform/kubernetes/issues/7018#issuecomment-99749177).

Prefixing ActiveDeadlineSeconds with "Node" is slightly less ambiguous, but at the cost of being more verbose. I'm not sure it's worth it, so long as the description string explains it adequately, especially if there's a matching Active condition paired with it.

bgrant0607 commented:
cc @dchen1107 @vmarmol

davidopp commented:
/subscibe

derekwaynecarr commented:
@bgrant0607 - thanks for feedback as always.

I will update this PR to the following:

1. PodStatus.StartTime - when a Kubelet acknowledges the pod is scheduled to its host, and begins to start pulling requisite images and run the container.
2. PodSpec.ActiveDeadlineSeconds - number of seconds relative to StartTime until Kubelet will actively work to kill the pod.

Once we settle on *Time fields, and representation of Duration fields I would be happy to update in a follow-on.  ActiveDeadlineSeconds is at least consistent with the graceful termination field on PodSpec.  And when we look to eliminate Phase, I agree we should be able to map PodStatus.StartTime to a relevant condition.

Will ping when ready for another pass.

derekwaynecarr commented:
Having a little bit of trouble getting a ```PodStatus.StartTime``` that does not get reset every sync loop given the current Kubelet code structure.   Hope to get a work-around soon, but currently banging head.  All status now is derived from what Docker returns, and we are aggressive about deleting status if there is a batch failure.

bgrant0607 commented:
@vmarmol, @dchen1107, or someone else from the node team may have some suggestions.

dchen1107 commented on pkg/api/v1/types.go:
```diff
@@ -807,6 +807,7 @@ type PodSpec struct {
 	// a termination signal and the time when the processes are forcibly halted with a kill signal.
 	// Set this value longer than the expected cleanup time for your process.
 	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty" description:"optional duration in seconds the pod needs to terminate gracefully; may be decreased in delete request; value must be non-negative integer; the value zero indicates delete immediately; if this value is not set, the default grace period will be used instead; the grace period is the duration in second ...
+	RunningTtl                    *int64 `json:"runningTtl,omitempty" description:"optional duration in seconds the pod may run before the system will actively try to mark it failed and kill associated containers; value must be a positive integer`
```
validation on the value?

derekwaynecarr replied:
That is coming in update to PR

dchen1107 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1065,6 +1065,46 @@ func (kl *Kubelet) cleanupOrphanedVolumes(pods []*api.Pod, runningPods []*kubeco
+// filterOutPodsExceedRuntimeTtl filters pods that have a container whose startAt time is more
+// than the specified RunningTtl for the pod.  It records an event that the pod has run longer
+// than the allocated time, and updates the pod status as failed.  By filtering the pod from
+// the result set, the Kubelet will kill the pod's containers as part of normal SyncPods workflow
```
How this will be interactivate with restart policy -- RestartOnFailure and RestartAlways?

dchen1107 replied:
Answered by following implementation.

derekwaynecarr replied:
Goal is to kill pod and mark failed so it's independent of restart policy.  Issue is getting a good stable date and time to make the relative value.  It looked like taking the oldest container started at was insufficient because it did change if there ever was a container restart.  So in my local branch I tried to keep a StartTime aside from the container statuses in status manager.  Will update for review after I push an updated branch.

derekwaynecarr commented:
@dchen1107 I will update this PR over weekend with latest changes and re ping with feedback based on any problems I am having on Monday

derekwaynecarr commented:
Ok, got something working, will finish clean up and update PR.  Need to run some tests with what happens if Kubelet dies and comes back.

derekwaynecarr committed:
Kubelet kills a pod that exceeds ActiveDeadlineSeconds

derekwaynecarr commented:
Ok, this should be ready for a review pass now.  Tests added, swagger doc generated, etc.

davidopp commented on pkg/api/types.go:
```diff
@@ -835,6 +838,9 @@ type PodStatus struct {
 	HostIP string `json:"hostIP,omitempty"`
 	PodIP  string `json:"podIP,omitempty"`
 
+	// date and time at which the object was acknowledged by the Kubelet
```
I think you should expand a little on "acknowledged by the Kubelet." At least mention that this is before Kubelet tries to pull the container image.

davidopp commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -4205,3 +4205,110 @@ func TestFilterOutTerminatedPods(t *testing.T) {
+		},
+	}
+
+	// format is // k8s_<container-id>_<pod-fullname>_<pod-uid>
```
extraneous second //

davidopp commented:
LGTM but is there any way to add a test to verify that a pod that exceeds its deadline actually gets killed?

vmarmol commented on pkg/kubelet/status_manager.go:
```diff
@@ -74,6 +74,11 @@ func (s *statusManager) SetPodStatus(pod *api.Pod, status api.PodStatus) {
 	defer s.podStatusesLock.Unlock()
 	oldStatus, found := s.podStatuses[podFullName]
+
+	// we preserve the previous cached start time if present
```
Missing last part of this comment?

bgrant0607 commented:
Please document and add a validation check that the deadline cannot be set when restart policy is always.

derekwaynecarr commented:
Added unit tests to ensure that SyncPods will terminate the actual container if the deadline is exceeded, and that it will not terminate the pod if the deadline has not been reached.

Updated swagger documentation to provide more detail on the meaning of StartTime.

@bgrant0607 - I do not follow your last comment.  The restart policy only affects if the containers are restarted in light of a failure, but if the pod itself is failed, nothing should restart.  Right now, if a restart always pod exceeds it active deadline, the pod is killed and all containers are killed.  If prior to that deadline being reached, any of the individual containers of the pod were to fail, they are restarted as many times as needed up until the specified deadline.  This seems like the right behavior to me, let me know if I misunderstood something.

dchen1107 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1741,6 +1777,11 @@ func (kl *Kubelet) generatePodStatus(pod *api.Pod) (api.PodStatus, error) {
 		podStatus.HostIP = hostIP.String()
 	}
 
+	podStatus.StartTime = pod.Status.StartTime
```
Can we consolidate resetting StartTime logic into one place so that it is not that fragile due to lack of checkpointing in kubelet today? How about moving it together with SetPodStatus in status_manager?

derekwaynecarr replied:
Will move.

dchen1107 commented:
LGTM overall except one nit.

@bgrant0607 My understanding about this feature is that it is independent from restart policy completely. Or we don't agree on that?

thockin commented on pkg/api/types.go:
```diff
@@ -835,6 +838,10 @@ type PodStatus struct {
 	PodIP  string `json:"podIP,omitempty"`
 
+	// Date and time at which the object was acknowledged by the Kubelet.
+	// This is before the Kubelet pulled the container image(s) for the pod.
```
Is it right to include pull time in this deadline?

dchen1107 replied:
Ideally, we should separate the limit of pulling all images in a pod from how long a container in pod could run, but that is much more complicated today. Also I feel one bound limit to the end user is much easy to understand.

derekwaynecarr commented:
@dchen1107 - addressed your request to localize computation of start time to status_manager only.

derekwaynecarr commented:
Looks like signature for TestKubelet changed, fixing up.

derekwaynecarr commented:
Pushed an update to respond to change, Travis should hopefully show green now.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1159,6 +1192,7 @@ func (kl *Kubelet) SyncPods(allPods []*api.Pod, podSyncTypes map[types.UID]metri
 	//      though the pod was not considered terminated by the apiserver).
 	// These two conditions could be alleviated by checkpointing kubelet.
 	pods := kl.filterOutTerminatedPods(allPods)
+	pods = kl.filterOutPodsPastActiveDeadline(pods)
```
A nit: could you follow the pattern similar to `handleNotFittingPods` and `handleOutofDisk` functions above (line 1179/1182)? They set the pods status to fail and let `filterOutTerminatedPods` handles all the filtering.

derekwaynecarr replied:
ok, so just flip the order of the function calls?

derekwaynecarr replied:
I see, will update.

derekwaynecarr replied:
Hmm, I am not sure I like the proposed change.  It makes me have to filter out pods that are already terminated in the filter function which seems to duplicate the purpose of filterOutTerminatedPods

derekwaynecarr replied:
In fact, I am a little confused how the existing ``handle**``` functions are not causing duplicate events from being recorded for hostPortConflicts, etc.  Shouldn't those functions ignore failed pods as well?  Can a pod in a failure state actually cause a hostPortConflict right now if a new pod is scheduled?

yujuhong replied:
Uh...what I meant was that you only need to set the pod status to fail, and filterOutTerminatedPods would just handle the filtering. No need to duplicate code in your own function.

yujuhong replied:
To make that happen, you will have to set the pod status to fail in your function *before* kubelet calls filterOutTerminatedPods

derekwaynecarr replied:
I am now of the opinion that the proper behavior is to move the ```handle***``` functions to after the filter call to stop checking pods that are failed or succeeded in this routine and having that potentially change the state of a succeeded pod as a host port conflict or as not fitting.  At least, that is how my current read of the code makes me think what is happening.  Please clarify if I am missing something.

derekwaynecarr replied:
I guess what I find confusing about the current structure is the following:

Step 1: handleNotFittingPods
The input list of pods to this function is all pods bound to the Kubelet.  It checks the pod.Spec.Containers to check for host port conflicts.  If a pod is failed, it does not make sense that it should be taken into account for host port conflicts at all. If there is a conflict though, we fail a pod, and throw an event.

Step 2: Filter out the terminated pods

Step 3: Filter out the pods past active deadline
Since we already have a list of pods that were not terminated, we do not get called twice for the same pod.  This stops duplicate events.

... REPEAT ...

On the subsequent repeat, we throw a host port conflict again because we do not ignore terminated pods for not fitting pods.  That seems like a bug in the current structure.

yujuhong replied:
@derekwaynecarr, I see your point. It seems like that the `handle*` functions would cause duplicated events. They should each filters out terminated pods before processing. I will upload a PR to fix that, so you can leave your PR unchanged.

yujuhong replied:
FYI, #8274

derekwaynecarr commented:
I would like to rebase this PR, but it appears the code generator for conversions is not working for v1 api version

```
GOPATH=$(godep path):$GOPATH go run cmd/kube-conversion/conversion.go -v v1 -f file1.txt -n file2.txt
```

Both output files are empty.

derekwaynecarr commented:
Ok, fixed the error in kube-conversion.go as part of PR so I could generate v1 conversions.

Poking this PR so it can hopefully avoid further rebases.

derekwaynecarr commented:
This carries a commit from https://github.com/GoogleCloudPlatform/kubernetes/pull/8201

yujuhong commented on pkg/kubelet/status_manager.go:
```diff
@@ -74,6 +74,25 @@ func (s *statusManager) SetPodStatus(pod *api.Pod, status api.PodStatus) {
+			// based on last observed value
+			status.StartTime = pod.Status.StartTime
+		}
+	}
```
This changes status manager's behavior as it is no longer a dumb cache; `StartTime` cannot be re-computed by looking at the containers (`generatePodStatus`). I think it's fine but we should probably prevent others from bypassing the cache to get the status. I will do that after this PR lands.

yujuhong commented:
LGTM pending the rebase.

derekwaynecarr commented:
rebased, should be good to go.  integrated with your admitPods function.

yujuhong commented:
Shippable is green. Merging.
